### PR TITLE
[FIX] Dice loss computation in both backends

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- fix dice loss also for multiclass training

Any feedback is welcome :hugs:

### CODE REVIEW

**Comment by felixdittrich92** [comment]
@odulcy-mindee Failing tests will be fixed if #1436 is merged

File: doctr/models/detection/differentiable_binarization/pytorch.py [hunk 1]
```diff
@@ -263,11 +263,13 @@ def compute_loss(
             # Class reduced
             focal_loss = (seg_mask * focal_loss).sum() / seg_mask.sum()
 
-            # Compute dice loss for approx binary_map
+            # Compute dice loss for each class or for approx binary_map
             binary_map = 1 / (1 + torch.exp(-50.0 * (prob_map - thresh_map)))
-            inter = (seg_mask * binary_map * seg_target).sum()  # type: ignore[attr-defined]
-            cardinality = (seg_mask * (binary_map + seg_target)).sum()  # type: ignore[attr-defined]
-            dice_loss = 1 - 2 * (inter + eps) / (cardinality + eps)
+            dice_map = torch.softmax(out_map, dim=1) if len(self.class_names) > 1 else binary_map
```
Comment (odulcy-mindee) [inline doctr/models/detection/differentiable_binarization/pytorch.py:268]: For readibility and efficiency, you can:
```python
if len(self.class_names) > 1:
    dice_map = torch.softmax(out_map, dim=1)
else:
    # compute binary map instead
    dice_map = 1 / (1 + torch.exp(-50.0 * (prob_map - thresh_map)))
```
so you can remove `binary_map = 1 / (1 + torch.exp(-50.0 * (prob_map - thresh_map)))`

**Review by odulcy-mindee — COMMENTED** [review]
and same comments for the other files

**Review by odulcy-mindee — APPROVED** [review]

**Comment by felixdittrich92** [comment]
@odulcy-mindee one mypy issue we can fix this next week
